### ISSUE TITLE
Incorrect focus management when screen reader users click links within the megamenu & land on pages with left nav menu

### ISSUE THREAD

**Issue by laflannery** [department-of-veterans-affairs/va.gov-cms#10699]
## Describe the defect
On certain pages on VA.gov, if a screen reader user clicks on a link in the desktop megamenu, the focus remains in the megamenu instead of moving to the title of the content page. This creates confusion and frustration for the user.

## To Reproduce
(With your screen reader turned on)
1. Go to Va.gov [homepage](https://www.va.gov/)
2. Click on "VA Benefits and Health Care" in the header
3. Click on "Education and training" in the megamenu
4. Click on "Eligibility"
5. When the Eligibility page loads, the screen reader continues to read the links within the megamenu instead of the Eligibility page content.

## AC / Expected behavior

- [ ] After activating any link within the megamenu, focus should always be on the skip link at the top of the page
- [ ] Ensure that your solution works across VAgov, not selectively
- [ ] Confirm with the Accessibility Lead
- [ ] **After merge to prod,** validate whether your solution works on injected-header sites
- [ ] (if necessary) Create any follow-up tickets

 ## Additional context
After testing, I was not able to reproduce this behavior on all pages, only some.  I am unsure if there is something different about the pages, or the megamenu or what exactly is making this behavior inconsistent across various pages.

**Comment by apisandipas** [comment]
## Developer Notes 

### Recreation Attempt

When attempting to recreate the issue, using a screen reader, I've been unsuccessful when using the steps described above. Here is what I tried and experienced: 

(With screen reader turned on)
1. Go to Va.gov [homepage](https://www.va.gov/)
2. Click on "VA Benefits and Health Care" in the header
3. Click on "Education and training" in the megamenu
4. Click on "Eligibility"
5. When the Eligibility page loads, the screen reader reads the Skip to Content Link and when thats clicked, it started to read the links in the Left Side Navigation Links. 

Screen Reader used: 
- [Chrome Screen Reader](https://chrome.google.com/webstore/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn)
-  MacOS VoiceOver utility.

### Determination
While I could not reproduce the issue as described, I think we may be talking about the same thing after all. The links in the Left Side Navigation are nearly identical to those that appear in the Education section of the Mega Menu.

This makes sense given that in the markup, the parent element of both this Navigation and the main content area possesses the  "content" ID attribute. So it's behaving as expected, but probably not as desired. This isn't then a Mega Menu issue after all, but likely a template structure issue. It's likely that we may need to change all Templates with Left had navigation to an Aside Element for the navigation and wrap the main content with a Main tag as well as give it the "content" ID attribute. Thought it may be sufficient enough to move the "content" ID to the parent element of the main content itself rather than change the semantic structure of the page. 

To restate the issue: The id="content" element wraps both the side navigation menu and the main content section. I believe ideally it would only wrap the main content of the page, this given the "Skip To Content" link the desired effect of skipping to the main content section.

I would suggest we do a Spike ticket to determine LOE for making this change across the page templates. At first glance it appears the id="content" is in 49 different layout templates.

@laflannery - Can you confirm that what I've said is accurate in regards to the expected behavior?
@wesrowe @jilladams - Should this be made into a Spike ticket to make LOE determinations?

**Comment by wesrowe** [comment]
@laflannery @apisandipas - I just tried the Chrome screen reader extension and had a different experience: it read the H1 but the put focus on the VA logo (not even the skip link). 

This situation would benefit from a quick meeting to get on the same page and run Bryan's hypothesis by Laura. I'll try to find us a time.

**Comment by apisandipas** [comment]
I spoke with @laflannery on Zoom about the issue and we both agree that the behavior is too inconsistent to guess what what might be happening versus what ought to be happening. We plan to confer with a native screen-reader user to decided what should be happening and then determine a course of action from there.

**Comment by laflannery** [comment]
@apisandipas and @wesrowe Notes from my discussion with Angela:

- She agrees that it's reading the left nav - apparently she has mentioned this duplication of links between the megamenu and left navigation before and how it might be causing confusion and it seems that was the case here.
- Her recommendation is to force focus on page load to the H1:
  - This would be on content pages only.  For example we were looking at a page with a form, we would want focus to go to the first form field
  - Our previous concern about what gets skipped when focus is forced - Angela says this is ok.  Assistive technology users can easily go back to the left nav or the mega menu if you need/want to.  So she was not concerned about this
  - She compared it to a sighted user - as a sighted user what do you do first when a page loads? You look at and read the H1. Then if needed you scan the other parts of the page.  Forcing focus to the H1 will effectively do the same thing for a screen reader user.

Let me know if you have other concerns that I forgot to ask Angela, also if the solution we are asking for is not possible. Thanks!

**Comment by apisandipas** [comment]
@laflannery I want to make sure I'm clear on the ask here. Are we thinking that we want to have the H1 be the first focusable item on the pages with text content? Is this instead of the Skip to Content link? Or have it be the first thing focused *after* the skip to content link. I'm pretty sure I can force this by adding adding tabindex to the H1. I'm not sure if I would need to add on to the Skip Link as well. Perhaos tabindex=1 on the Skip Link and tabindex=2 on the H1 would be enough to force the order explicitly.

**Comment by laflannery** [comment]
@apisandipas The focus would be on the H1 instead of the skip link.  It's similar to the [search results pages](https://www.va.gov/search/?page=2&query=veterans&t=false). When I use the pagination, the focus is forced to the number of results "Showing 11-20 of 999 results for 'veterans'".

The issue with simply adding a tabindex is that then the H1 would be tabbable as a non-interactive element, which is not what we want.  Again, you can look at that search results page - the results number gets focus when the page loads, but if you use the keyboard to tab, you can't actually focus on/tab to that text.

Does that help explain how we ideally want this to work?  Let me know if you want to jump on another call.

**Comment by jilladams** [comment]
Discussed this question with @wesrowe . This is a big change that will affect all screen reader users on all content pages (with left sidebar? or without?), we need to pause and evaluate a bit further before we implement. 

Next step: Wes (& Jill) to chat with Dave about it and understand if we need further user testing or how best to proceed. For now, at min, rewriting issue title to better reflect where we've landed.

**Comment by jilladams** [comment]
Dave's feedback:
* PW to audit / groom: how many pages are affected by this left nav experience? Only benefit detail pages / any other templates to consider?
* Laura is the final feedback to inform Sitewide accessibility decisions

@laflannery can you clarify the nature of this being a defect 2? Dave asked that, and we just want to confirm if that stands based on what we know now.

**Comment by laflannery** [comment]
@jilladams Yes defect-2 still applies, the user is still not getting the primary H1 first which is the issue so it causes confusion as to where exactly they are

**Comment by jilladams** [comment]
^ made #11347 . I suspect Bryan partially knows the answer, but will close loop when he's back online this week to confirm if we need to treat that as an injection / how effortful to answer the questions there.

**Comment by laflannery** [comment]
@jilladams FYI Dave and I spoke about this today - he wanted to talk about this more holistically rather than just our team fixing just this one template/issue.  With that in mind I have a meeting set up tomorrow with some of the a11y people on the platform team to discuss and anything that comes out of that I will let you know.

**Comment by jilladams** [comment]
@laflannery I lost track of the follow up from the a11y meeting about this. Could you remind re: any outcomes of that meeting?

**Comment by laflannery** [comment]
@jilladams Some notes from that meeting, unfortunately there is no clear answer as to what definitely should be done:
- We do not have an already established pattern for focus management in this context, meaning it would have to be established for our scenario and how it should work throughout the site
- We do not have a clear answer as to why there does not seem to be consistency on what order page elements are read by a screen reader when a page is loaded
- It was suggested that perhaps it was related to page performance.  I did ask Bryan and he isn't sure this is the culprit.
- I am trying to do some additional research to see if I can find a pattern/reason for why pages with the same template would not have the same behavior for a screen reader user, whether it be page performance or otherwise.
  - If the page performance is discovered to be impacting/causing the behavior and confusion, focus management is most likely the way to resolve this.
- Another suggestion was perhaps the page template itself may need to be looked at, to make it clearer, simpler and easier to be understood by a screen reader.

**Comment by jilladams** [comment]
Helpful thanks - were there any notes re: why folks think performance may be implicated?

**Comment by jilladams** [comment]
Per sync with Dave: 
* @laflannery to provide examples of pages that are similar content type & presumably same template, in order for PW to review in code & confirm if any code differences are contributing.
* Laura to review process with Platform to determine documented rules for how to handle focus on the page

**Comment by laflannery** [comment]
The pages I used to test we the pages in the first column of the mega menu options:
![image](https://user-images.githubusercontent.com/106678594/199501146-cadb92c8-b51a-4982-aeab-3706a540a756.png)
![image](https://user-images.githubusercontent.com/106678594/199501235-ba640480-a323-45ec-bafa-d4506babf495.png)

Using a screen reader, the majority of pages seemed to follow this pattern:

- The first time I went to the page, the SR would finish reading the mega menu list I came from, then it would read the new page in the following order:
  - `<title>` tag
  - Skip link
  - Pact act
  - Breadcrumbs
  - Left nav, starting with the H4 heading
- Subsequent times I went to the same page however, the pages would not be read in the same way, they would use one of two patterns:
  1. The `<title>` tag would be read first followed by the Left Nav without the H4 heading OR
  2. The `<title>` tag would be read then the SR would stop

Examples of pages that followed scenario 1 the second time I went to the page:
- https://www.va.gov/health-care/about-va-health-benefits/
- https://www.va.gov/health-care/how-to-apply/
- https://www.va.gov/burials-memorials/eligibility/
- https://www.va.gov/careers-employment/vocational-rehabilitation/

Examples of pages that followed scenario 2 the second time I sent to the page:
- https://www.va.gov/education/eligibility/
- https://www.va.gov/disability/how-to-file-claim/

**Comment by wesrowe** [comment]
@laflannery, thanks for the deeper dive in trying to reproduce the behaviors consistently. I have written down two action items for this from our sync yesterday:
- Have a FE engineer try to figure out the differences in code/content that are driving the divergent behaviors you outline 
- Try to get the Platform Accessibility crew to provide a pattern we should be following in any given scenario

I'll chat with @jilladams and @apisandipas about whether to create a new ticket for the first item.

I assume you'll do the second bullet – how do you want to track it?

**Comment by laflannery** [comment]
@Wes I know the a11y specialist working with the DST, I believe this would be best to ask her and raise with her first so I have done that.  If she tells me that's not the right team I'll have to see where next to go but let me see what she says because I believe she can give me direction (i.e. create a ticket for DST for us to follow with them)

**Comment by laflannery** [comment]
@wesrowe and @jilladams Regarding where the focus should be placed on page load, this is from the accessibility specialist on the platform design team, Sarah Koomson:

> Focus should always start at the top of the page, so whatever element is first. Usually an accessibility link (if an accessibility version is available) or a skip to main content link.

https://dsva.slack.com/archives/C01DBGX4P45/p1667581173335559?thread_ts=1667501506.319509&cid=C01DBGX4P45

**Comment by jilladams** [comment]
Moving back to Unrefined while we figure out next steps. (cc @wesrowe )

**Comment by jilladams** [comment]
Next step: #11347 
https://dsva.slack.com/archives/C52CL1PKQ/p1667933509460979?thread_ts=1667852808.205269&cid=C52CL1PKQ

**Comment by jilladams** [comment]
#11347 was deemed low priority til after the federal header update ships, and was iceboxed. Iceboxing this with it.

**Comment by FranECross** [comment]
Per Dave Conlon, the Federal Header/Footer improvement effort has been closed by the Whitehouse / 18f. All related tickets may be closed.

Because of this, I'll be looking at some of the tickets ice-boxed to determine if they are again viable in light of the above.

**Comment by FranECross** [comment]
Per Dave Conlon, the Federal Header/Footer improvement effort has been closed by the Whitehouse / 18f. All related tickets may be closed. The DST work to componentize is no longer predicated on this.

**Comment by randimays** [comment]
@FranECross Should this be transferred to DST (as owners of the header)?

**Comment by FranECross** [comment]
@randimays Yes, thank you for flagging this.

**Comment by FranECross** [comment]
Closing old ticket. DST will create tickets as needed (they own the header/footer now).
